File: packages/openland-graphql/spacex/transport/net/WatchDogTimer.ts
export class WatchDogTimer {

    readonly onRestart: () => void;
    readonly timeout: number;
    private timer: any;
    private isDead = true;

    constructor(timeout: number, onRestart: () => void) {
        this.timeout = timeout;
        this.onRestart = onRestart;
    }

    kick = () => {
        if (this.isDead) {
            return;
        }
        if (this.timer) {
            clearTimeout(this.timer);
        }
        this.timer = setTimeout(() => {
            if (!this.isDead) {
                this.isDead = true;
                this.onRestart();
            }
        }, this.timeout);
    }

    reset = () => {
        this.isDead = false;
        this.kick();
    }

    kill = () => {
        this.isDead = true;
        if (this.timer) {
            clearTimeout(this.timer);
            this.timer = null;
        }
    }
}

File: android/app/src/main/java/com/openland/spacex/transport/net/WatchDogTimer.kt
package com.openland.spacex.transport.net

import com.openland.spacex.utils.DispatchQueue

class WatchDogTimer(private val timeout: Int, private val onRestart: () -> Unit, private val queue: DispatchQueue) {

    private var isDead = true
    private var timer: (() -> Unit)? = null

    fun kick() {
        if (!this.queue.isCurrentQueue) {
            throw Error("Invalid Queue")
        }
        if (this.isDead) {
            return
        }
        this.timer?.invoke()
        this.timer = null

        this.timer = this.queue.asyncDelayed(this.timeout) {
            this.timer = null
            if (!this.isDead) {
                this.isDead = true
                this.onRestart()
            }
        }
    }

    fun reset() {
        if (!this.queue.isCurrentQueue) {
            throw Error("Invalid Queue")
        }
        this.isDead = false
        this.kick()
    }

    fun kill() {
        if (!this.queue.isCurrentQueue) {
            throw Error("Invalid Queue")
        }

        this.isDead = true
        this.timer?.invoke()
        this.timer = null
    }
}